### ISSUE TITLE
[Bug] CGO target headers not added to search paths when imported as external repository

### ISSUE THREAD

**Issue by Nickersoft** [bazelbuild/rules_go#3109]
### What version of rules_go are you using?
0.31.0

### What version of gazelle are you using?
0.24.0

### What version of Bazel are you using?
5.1.0

### Does this issue reproduce with the latest releases of all the above?
Yes, technically (I can't use Gazelle 0.25.0 due to [this issue](https://github.com/bazelbuild/bazel-gazelle/issues/1217))

### What operating system and processor architecture are you using?
macOS Monterey 12.3.1 (21E258)
2.7 GHz Quad-Core Intel Core i7

### Any other potentially useful information about your toolchain?
Using Go 1.18

### What did you do?
I'm currently trying to import a Bazel project that utilizes CGO + JNI (available [here](https://github.com/theopendictionary/odict/tree/nickersoft/bazel5)) into another repository (available [here](https://github.com/nickersoft/go-bazel-repro)).

### What did you expect to see?
I expected it to work as usual.

### What did you see instead?
While the [original repo](https://github.com/theopendictionary/odict/tree/nickersoft/bazel5) compiles fine, trying to import it into another project does not add the correct header search paths to the command line, despite the JNI target (`//java/main/cpp`) directly requiring the CGO target (`//bridge`).

The CGO target inside `//bridge` produces a `bridge.h` file that is imported as `bridge/bridge.h` in the [odict.cpp](https://github.com/TheOpenDictionary/odict/blob/master/java/main/cpp/odict.cpp) file. Inside the original repo, the `bazel build java` command executes the following in the sandbox:

```
cd /private/var/tmp/_bazel_tjnickerson/1bdab46624ea7f4407b2d453e6808fb8/sandbox/darwin-sandbox/232/execroot/odict && \
  exec env - \
    APPLE_SDK_PLATFORM=MacOSX \
    APPLE_SDK_VERSION_OVERRIDE=12.3 \
    DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer \
    PATH=/Users/tjnickerson/Library/Caches/bazelisk/downloads/bazelbuild/bazel-5.1.1-darwin-x86_64/bin:/Users/tjnickerson/.sdkman/candidates/java/current/bin:/usr/local/opt/ruby@2.7/bin:/Users/tjnickerson/Library/Caches/fnm_multishells/9481_1649548540226/bin:/usr/local/anaconda3/bin/:/Users/tjnickerson/go/bin:/Users/tjnickerson/base/phacility/arcanist/bin:/Users/tjnickerson/.yarn/bin:/Users/tjnickerson/.config/yarn/global/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/tjnickerson/.rvm/bin \
    SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk \
    TMPDIR=/var/folders/0z/yzv926zn0fzbtjjw3p_w0s2r0000gn/T/ \
    XCODE_VERSION_OVERRIDE=13.3.0.13E113 \
    ZERO_AR_DATE=1 \
  /usr/bin/sandbox-exec -f /private/var/tmp/_bazel_tjnickerson/1bdab46624ea7f4407b2d453e6808fb8/sandbox/darwin-sandbox/232/sandbox.sb /var/tmp/_bazel_tjnickerson/install/7a9606eb8f092191f2180bf868dc1b49/process-wrapper '--timeout=0' '--kill_delay=15' external/local_config_cc/wrapped_clang_pp '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -fno-omit-frame-pointer -O0 -DDEBUG '-std=c++11' 'DEBUG_PREFIX_MAP_PWD=.' -iquote . -iquote bazel-out/darwin-fastbuild/bin -Ibazel-out/darwin-fastbuild-ST-97ca104753ae/bin -isystem tools/jni -isystem bazel-out/darwin-fastbuild/bin/tools/jni -MD -MF bazel-out/darwin-fastbuild/bin/java/main/cpp/_objs/odict-lib/odict.d '-frandom-seed=bazel-out/darwin-fastbuild/bin/java/main/cpp/_objs/odict-lib/odict.o' -isysroot __BAZEL_XCODE_SDKROOT__ -F__BAZEL_XCODE_SDKROOT__/System/Library/Frameworks -F__BAZEL_XCODE_DEVELOPER_DIR__/Platforms/MacOSX.platform/Developer/Library/Frameworks '-mmacosx-version-min=12.3' -no-canonical-prefixes -pthread -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -target x86_64-apple-macosx -c java/main/cpp/odict.cpp -o bazel-out/darwin-fastbuild/bin/java/main/cpp/_objs/odict-lib/odict.o
```

which includes the following header search paths:

- `.`
- `bazel-out/darwin-fastbuild/bin` 
- `bazel-out/darwin-fastbuild-ST-97ca104753ae/bin` (`bridge/bridge.h` lives here, so the import works as expected)
-  `tools/jni` 
- `bazel-out/darwin-fastbuild/bin/tools/jni`

yet running `bazel build main` on my [reproduction repo](https://github.com/nickersoft/go-bazel-repro) runs this in the sandbox:

```
cd /private/var/tmp/_bazel_tjnickerson/8b9d21129a210b6f322b695a12a00caf/sandbox/darwin-sandbox/142/execroot/__main__ && \
  exec env - \
    APPLE_SDK_PLATFORM=MacOSX \
    APPLE_SDK_VERSION_OVERRIDE=12.3 \
    DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer \
    PATH=/Users/tjnickerson/Library/Caches/bazelisk/downloads/bazelbuild/bazel-5.1.1-darwin-x86_64/bin:/Users/tjnickerson/.sdkman/candidates/java/current/bin:/usr/local/opt/ruby@2.7/bin:/Users/tjnickerson/Library/Caches/fnm_multishells/14807_1649549587978/bin:/usr/local/anaconda3/bin/:/Users/tjnickerson/go/bin:/Users/tjnickerson/base/phacility/arcanist/bin:/Users/tjnickerson/.yarn/bin:/Users/tjnickerson/.config/yarn/global/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/tjnickerson/.rvm/bin \
    SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk \
    TMPDIR=/var/folders/0z/yzv926zn0fzbtjjw3p_w0s2r0000gn/T/ \
    XCODE_VERSION_OVERRIDE=13.3.0.13E113 \
    ZERO_AR_DATE=1 \
  /usr/bin/sandbox-exec -f /private/var/tmp/_bazel_tjnickerson/8b9d21129a210b6f322b695a12a00caf/sandbox/darwin-sandbox/142/sandbox.sb /var/tmp/_bazel_tjnickerson/install/7a9606eb8f092191f2180bf868dc1b49/process-wrapper '--timeout=0' '--kill_delay=15' external/local_config_cc/wrapped_clang_pp '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -fno-omit-frame-pointer -O0 -DDEBUG '-std=c++11' 'DEBUG_PREFIX_MAP_PWD=.' -iquote external/odict -iquote bazel-out/darwin-fastbuild/bin/external/odict -Ibazel-out/darwin-fastbuild-ST-97ca104753ae/bin -isystem external/odict/tools/jni -isystem bazel-out/darwin-fastbuild/bin/external/odict/tools/jni -MD -MF bazel-out/darwin-fastbuild/bin/external/odict/java/main/cpp/_objs/odict-lib/odict.d '-frandom-seed=bazel-out/darwin-fastbuild/bin/external/odict/java/main/cpp/_objs/odict-lib/odict.o' -isysroot __BAZEL_XCODE_SDKROOT__ -F__BAZEL_XCODE_SDKROOT__/System/Library/Frameworks -F__BAZEL_XCODE_DEVELOPER_DIR__/Platforms/MacOSX.platform/Developer/Library/Frameworks '-mmacosx-version-min=12.3' -no-canonical-prefixes -pthread -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -target x86_64-apple-macosx -c external/odict/java/main/cpp/odict.cpp -o bazel-out/darwin-fastbuild/bin/external/odict/java/main/cpp/_objs/odict-lib/odict.o
```

which uses the following header search paths:

- `external/odict` 
- `bazel-out/darwin-fastbuild/bin/external/odict` (only includes `java`, `schema`, and `tools` – would expect it to contain `bridge` too seeing it's required by the build)
- `bazel-out/darwin-fastbuild-ST-97ca104753ae/bin` (`bridge.h` now technically lives under `bazel-out/darwin-fastbuild-ST-97ca104753ae/bin/external/odict/bridge/bridge.h`, and I have no way of accessing it without changing the original import to `#include "external/odict/bridge/bridge.h"`, which would break the build in the original repo)
- `external/odict/tools/jni`
- `bazel-out/darwin-fastbuild/bin/external/odict/tools/jni`

and throws the error:

```
external/odict/java/main/cpp/odict.cpp:4:10: fatal error: 'bridge/bridge.h' file not found
#include "bridge/bridge.h"
         ^~~~~~~~~~~~~~~~~
1 error generated.
```

This error seems to have cropped up out of nowhere, but I'm wondering if it might be the CGO target's failure to notify Bazel of its generated header or something? Also, if this is a larger-scale Bazel issue, I'm happy to report it to another repo, though I thought seeing the header is originating from CGO this would be a good place to start. Any help here would be greatly appreciated!

**TLDR:** to reproduce, clone [this repo](https://github.com/nickersoft/go-bazel-repro) and run `bazel build main` using Bazel 5.1.0. Should error.

**Comment by fmeum** [comment]
@Nickersoft Could you check whether #3115 fixes the issue? I successfully tested it on your reproducer both with and without `--experimental_sibling_repository_layout`.

**Comment by fmeum** [comment]
@Nickersoft Could you try again with this diff applied to your reproducer?

```
diff --git a/WORKSPACE b/WORKSPACE
index 436bf95..a3fb85d 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -9,6 +9,13 @@ http_archive(
     url = "https://github.com/TheOpenDictionary/odict/archive/%s.tar.gz" % ODICT_VERSION,
 )
 
+http_archive(
+    name = "io_bazel_rules_go",
+    sha256 = "5094563ee75463cdb600e32f8161a1366636591988c7cda9b8d1da3ca2cd67bc",
+    strip_prefix = "rules_go-76c459cad6fbe83f3db60443a28dd58202b63be9",
+    url = "https://github.com/fmeum/rules_gob/archive/76c459cad6fbe83f3db60443a28dd58202b63be9.tar.gz",
+)
+
 load("@odict//bazel:odict_deps.bzl", "odict_deps")
 
 odict_deps()
```

I'm hitting an issue with an `#include "bridge/archive.h"` that should be an `#include "bridge/bridge.h"`, but with that change, everything works.

**Comment by Nickersoft** [comment]
Ah yes, the `bridge` typo is an error with the upstream I've since fixed. But yeah, I just tested with my reproduction repo and my own personal Java codebase and everything works!! Thanks so much for this change :) Looking forward to seeing it in the next release.

Out of curiosity, do you know what recent changes could have caused this issue out of nowhere? Were there any major Bazel changes that affected this?

**Comment by fmeum** [comment]
I didn't look into that in detail, but Bazel 5 started using transitions in more places (e.g. genrule tools). It's possible that previous versions of Bazel picked up the missing include path from another dependency that is now built in a different configuration and thus lives in a separate directory.

**Comment by Nickersoft** [comment]
Ah, that would make sense.. I also didn't know Bazel had a `--experimental_sibling_repository_layout` for changing the way the external repos are organized. Hopefully as Bazel iterates there isn't another regression that breaks this haha.

I'll use a reference to your fork until your PR lands. Thanks again for all the help!!

**Comment by fmeum** [comment]
@Nickersoft Could you reopen the issue? It would make sense to keep it open until the fix has been reviewed and merged.
